### PR TITLE
Feature/hal pwm 2

### PR DESCRIPTION
**Pull request by RobertoAldea**

### CODE REVIEW

File: firmware/Sources/HAL/HAL_PWM/hal_pwm.c [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+/**********************************************************************************/
+/*                         Definition of local variables                          */
+/**********************************************************************************/
+static HRTIM_CompareCfgTypeDef pCompareCfg = {0};
+/**********************************************************************************/
+/*                        Definition of exported variables                        */
+/**********************************************************************************/
+extern HRTIM_HandleTypeDef hhrtim1;
+
```
Comment (pastorpflores) [inline firmware/Sources/HAL/HAL_PWM/hal_pwm.c:42]: add ´extern uint8_t EPC_ST_ERR_COUNTER´

File: firmware/Sources/HAL/HAL_PWM/hal_pwm.c [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+/**********************************************************************************/
+/*                         Definition of local functions                          */
+/**********************************************************************************/
+
+/**********************************************************************************/
+/*                        Definition of exported functions                        */
+/**********************************************************************************/
+
+HAL_PWM_result_e HAL_PwmInit(void){
+	HAL_PWM_result_e res = HAL_PWM_RESULT_SUCCESS;
```
Comment (pastorpflores) [inline firmware/Sources/HAL/HAL_PWM/hal_pwm.c:64]: change logict to init with HAL_PWM_RESULT_ERROR

File: firmware/Sources/HAL/HAL_PWM/hal_pwm.c [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+		res = HAL_PWM_RESULT_ERROR;
+	}
+	return res;
+}
+
+
+HAL_PWM_result_e HAL_PwmSetDuty(const uint32_t duty){
+	uint32_t max = hhrtim1.Instance->sTimerxRegs[0].PERxR;
+	uint32_t pwm_duty;
+	if (duty>100000){
```
Comment (pastorpflores) [inline firmware/Sources/HAL/HAL_PWM/hal_pwm.c:77]: no magic numbers. 

add internal #define to set constants

File: firmware/Sources/HAL/HAL_PWM/hal_pwm.c [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+		pwm_duty = 9215;
+	}
+	else{
+		pwm_duty = duty * max/100000 - 1;
+	}
+	if (pwm_duty < 96 || pwm_duty>9215){
```
Comment (pastorpflores) [inline firmware/Sources/HAL/HAL_PWM/hal_pwm.c:83]: if too big you are asigning the minimum duty too. This is wrong.

File: firmware/Sources/HAL/HAL_PWM/hal_pwm.h [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+ * @brief Start the pwm output
+ *
+ * @return @ref HAL_PWM_RESULT_SUCCESS if the pwm has started correctly,
+ * @ref HAL_PWM_RESULT_ERROR otherwise.
+ */
+HAL_PWM_result_e HAL_PwmStart (void);
+
+
+/**
+ * @fn HAL_PWM_pin_value_e HAL_PwmStop (void)
```
Comment (pastorpflores) [inline firmware/Sources/HAL/HAL_PWM/hal_pwm.h:92]: wrong definition here too

File: firmware/Sources/Test/HAL/hal_pwm_test.c [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+/*                  Include common and project definition header                  */
+/**********************************************************************************/
+
+
+/**********************************************************************************/
+/*                        Include headers of the component                        */
+/**********************************************************************************/
+#include "hal_pwm_test.h"
+
+#include "main.h"
```
Comment (pastorpflores) [inline firmware/Sources/Test/HAL/hal_pwm_test.c:16]: not needed. dont include main never.

File: firmware/Sources/Test/HAL/hal_pwm_test.h [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+/**********************************************************************************/
+/*                    Declaration of local function prototypes                    */
+/**********************************************************************************/
+
+/**********************************************************************************/
+/*                    	 Declaration of exported functions                  	  */
+/**********************************************************************************/
+
+/**
+ * @fn void HAL_PwmTest()
```
Comment (pastorpflores) [inline firmware/Sources/Test/HAL/hal_pwm_test.h:19]: wrong definition

**Review by pastorpflores — CHANGES_REQUESTED** [review]

File: firmware/Sources/HAL/HAL_PWM/hal_pwm.c (hunk 1 above)
Comment (RobertoAldea) [reply firmware/Sources/HAL/HAL_PWM/hal_pwm.c:42]: Done

File: firmware/Sources/HAL/HAL_PWM/hal_pwm.c (hunk 2 above)
Comment (RobertoAldea) [reply firmware/Sources/HAL/HAL_PWM/hal_pwm.c]: Done

File: firmware/Sources/Test/HAL/hal_pwm_test.c (hunk 6 above)
Comment (RobertoAldea) [reply firmware/Sources/Test/HAL/hal_pwm_test.c]: Done

File: firmware/Sources/HAL/HAL_PWM/hal_pwm.h (hunk 5 above)
Comment (RobertoAldea) [reply firmware/Sources/HAL/HAL_PWM/hal_pwm.h]: Done

File: firmware/Sources/HAL/HAL_PWM/hal_pwm.c (hunk 3 above)
Comment (RobertoAldea) [reply firmware/Sources/HAL/HAL_PWM/hal_pwm.c]: Done

File: firmware/Sources/HAL/HAL_PWM/hal_pwm.c [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+/**********************************************************************************/
+
+/**********************************************************************************/
+/*                         Definition of local variables                          */
+/**********************************************************************************/
+static HRTIM_CompareCfgTypeDef pCompareCfg = {0};
+/**********************************************************************************/
+/*                        Definition of exported variables                        */
+/**********************************************************************************/
+extern HRTIM_HandleTypeDef hhrtim1;
```
Comment (pastorpflores) [inline firmware/Sources/HAL/HAL_PWM/hal_pwm.c:41]: This is an imported variable

File: firmware/Sources/HAL/HAL_PWM/hal_pwm.c [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+
+
+HAL_PWM_result_e HAL_PwmSetDuty(const uint32_t duty){
+	uint32_t max = hhrtim1.Instance->sTimerxRegs[0].PERxR;
+	uint32_t pwm_duty;
+	if (duty>MAX_DUTY){
+		pwm_duty = 9215;
+	}
+	//If duty is 0 no division is allow so assigned the final result directly
+	else if (duty==0){
```
Comment (pastorpflores) [inline firmware/Sources/HAL/HAL_PWM/hal_pwm.c:88]: duty <= MIN_DUTY (96), no?

File: firmware/Sources/HAL/HAL_PWM/hal_pwm.c [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+	}
+	else{
+		pwm_duty = duty * max/MAX_DUTY - 1;
+	}
+	// Minimum duty to apply to the pwm is 96 by hardware
+	if (pwm_duty < 96){
+		pwm_duty = 96;
+	}
+	pCompareCfg.CompareValue = pwm_duty;
+	HAL_PWM_result_e res = HAL_HRTIM_WaveformCompareConfig(&hhrtim1, HRTIM_TIMERINDEX_TIMER_A, HRTIM_COMPAREUNIT_1, &pCompareCfg);;
```
Comment (pastorpflores) [inline firmware/Sources/HAL/HAL_PWM/hal_pwm.c:99]: please, add comments.

**Review by pastorpflores — CHANGES_REQUESTED** [review]

File: firmware/Sources/HAL/HAL_PWM/hal_pwm.c (hunk 9 above)
Comment (RobertoAldea) [reply firmware/Sources/HAL/HAL_PWM/hal_pwm.c]: I changed the if structure

File: firmware/Sources/HAL/HAL_PWM/hal_pwm.c (hunk 8 above)
Comment (RobertoAldea) [reply firmware/Sources/HAL/HAL_PWM/hal_pwm.c]: Changed

File: firmware/Sources/HAL/HAL_PWM/hal_pwm.c [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+
+/**********************************************************************************/
+/*                        Definition of exported functions                        */
+/**********************************************************************************/
+
+HAL_PWM_result_e HAL_PwmInit(void){
+	HAL_PWM_result_e res = HAL_PWM_RESULT_ERROR;
+	EPC_ST_ERR_COUNTER = 0;
+	MX_HRTIM1_Init();
+	if (EPC_ST_ERR_COUNTER){
```
[the remark this answers is not in the document]
Comment (RobertoAldea) [reply firmware/Sources/HAL/HAL_PWM/hal_pwm.c]: Changed the structure

**Review by pastorpflores — APPROVED** [review]

**Pull request merged**